icu4c/source/i18n/unicode/curramt.h

```diff
@@ -121,7 +121,7 @@ class U_I18N_API CurrencyAmount: public Measure {
 };
 
 inline const CurrencyUnit& CurrencyAmount::getCurrency() const {
-    return (const CurrencyUnit&) getUnit();
+    return dynamic_cast<const CurrencyUnit&>(getUnit());
```

Comment:
I do not know how to deal with this dynamic_cast

Reply:
If I change this line to static_cast<> it will break the following test under CFI
format/NumberFormatTest/TestCases
with the following run time error 
```
=== Handling test: format/NumberFormatTest/TestCases: ===
   format {
      NumberFormatTest {
         TestCases {
../../i18n/unicode/curramt.h:124:12: runtime error: control flow integrity check for type 'icu::CurrencyUnit' failed during base-to-derived cast (vtable address 0x7ff01c0bd5f0)
0x7ff01c0bd5f0: note: invalid vtable
 f0 7f 00 00  60 35 ef 1b f0 7f 00 00  90 34 ef 1b f0 7f 00 00  b0 34 ef 1b f0 7f 00 00  c0 34 ef 1b
              ^ 
../../i18n/unicode/curramt.h:124:12: note: check failed in /usr/local/google/home/ftang/icu/icu/icu4c/source/test/intltest/intltest, vtable located in ../../lib/libicui18n.so.73
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../../i18n/unicode/curramt.h:124:12 in 

```

Reply:
looks like it choke on line 77 of test/testdata/NumberFormatTestCases.txt
```
fpc: "en_US"        1234.56/USD  "$1,234.56"    1234.56/USD
```

Reply:
The line trigger such vtable issue is 
```
                     assertEquals(where + "getCurrencyFormat(" + mloc + ").parse(\"" + str + "\")",
                                  n, m);
```
in line source/test/intltest/numfmtst.cpp

Reply:
somehow if I move it from inline into .cpp it work with static_cast<>